### PR TITLE
Adds sleep immune trait and fixes latte

### PR DESCRIPTION
**Pull request by yorii**
:cl: Sleepimmunity Trait
add: Adds TRAIT_SLEEPIMMUNE which will make you immune to any and all sleep effects.
fix: Replaces SetSleeping(0) on cafe latte and soy latte with the new sleepimmunity trait.
/:cl:

[why]: Cafe Latte never had the intended effect of making you not sleep, it only resulted in popping you in and out of sleep every cycle if you had a toxin in your system putting you to sleep.

### CODE REVIEW

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm [hunk 1]
```diff
@@ -576,10 +576,17 @@
 	glass_name = "soy latte"
 	glass_desc = "A nice and refreshing beverage while you're reading."
 
+/datum/reagent/consumable/soy_latte/on_mob_add(mob/living/M)
+	M.add_trait(TRAIT_SLEEPIMMUNE)
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/drink_reagents.dm]: but what if you're an elf

Comment (yorii) [reply code/modules/reagents/chemistry/reagents/drink_reagents.dm]: Is there a joke I'm missing here? What about elves and cafe latte?

Comment (ExcessiveUseOfCobblestone) [reply code/modules/reagents/chemistry/reagents/drink_reagents.dm]: space elves next pr

**Comment by yorii** [comment]
Just realized this will conflict with #34893.
Both these PRs add the same trait but implement different code for handling it.

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm (hunk 1 above)
Comment (coiax) [reply code/modules/reagents/chemistry/reagents/drink_reagents.dm]: only if eating an elf corpse gives you sleep resistance, a la nethack

**Comment by XDTM** [comment]
The issue with checking for sleepimmune in the status effect is that you cannot bypass it if you need to for some future reason. It's better to check it in the mob proc itself in my opinion.

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm (hunk 1 above)
Comment (XDTM) [reply code/modules/reagents/chemistry/reagents/drink_reagents.dm]: you need to specify the source of the trait to add one, and it needs to be unique so it doesn't get overridden by another source

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm [hunk 2]
```diff
@@ -576,10 +576,17 @@
...
+	..()
+
+/datum/reagent/consumable/soy_latte/on_mob_delete(mob/living/M)
+	M.remove_trait(TRAIT_SLEEPIMMUNE)
```
Comment (PKPenguin321) [inline code/modules/reagents/chemistry/reagents/drink_reagents.dm]: But what if this is added by a different effect 🤔

Comment (XDTM) [reply code/modules/reagents/chemistry/reagents/drink_reagents.dm]: The proc needs an unique id as a second argument exactly because of that

**Comment by vuonojenmustaturska** [comment]
sleepimmune added in #34893, please ping a maintainer on irc or discord if you need this reopened

**Pull request closed without merging**